avalanche/training/plugins/checkpoint.py

```diff
@@ -212,6 +215,10 @@ def load_checkpoint_if_exists(
         return strategy, exp_counter
 
     def after_eval(self, strategy: BaseSGDTemplate, *args, **kwargs):
+        if self._training:
```

Comment:
isn't it easier to checkpoint inside `after_training_exp`?

Reply:
It is doable. In fact, we may checkpoint both after train and eval. In that case, we also need add to a flag to tell if the loaded checkpoint was created after train or eval (and add a couple of 'if's to the main script).

Reply:
is there any difference? no plugin should ever modify things inside the eval loop. I'm suggesting `after_training_exp` only because it seems the logical place to put it. If this is not true, maybe there is some complicated interaction that I'm ignoring.

Reply:
It's mostly related to how things are handled in the main script.

By checkpointing after eval, one can resume a checkpoint by just changing the initial experience like this:

```
for exp in train_steam[resumed_from:]:
  train()
  eval()
```

by checkpointing after train, you will end up with something like this:

```
resumed_from_checkpoint = True/False

for exp in train_steam[resumed_from-1:]:
  if not resumed_from_chechpoint:
    train()
  eval()
  resumed_from_checkpoint = False
```
(actually, we can simplify `resumed_from-1` by re-loading an already `-1`ed value.)

By checkpoint on both, things are actually not bad:

```
resume_from = 'eval/train' 
# resumed_from_checkpoint is not needed as we can just set resume_from='train' 
# same for the value of resumed_from, which would be adjusted accordingly

for exp in train_steam[resumed_from:]:
  if resume_from == 'train':
    train()
  eval()
```

Reply:
The assumption that you have an `eval` call is not necessarily true. If I call `strategy.train(exp, eval_streams=[valid_stream])` I may not need an additional eval since it's already inside the training loop. In fact, I could even have a single train call `strategy.train(train_stream, eval_streams=[train_stream, valid_stream])`.

However, we can safely do multiple `eval` calls. So, I would do something like this:

```
if resume_from_checkpoint:
    # load checkpoint here
    # also do eval, because we don't know if it was computed
    # this may not needed if the periodiceval does the final evaluation
    strategy.eval(valid_stream)

# we may or may not have a loop or an eval call here 
strategy.train(train_stream[resumed_from:], eval_streams=[train_stream, valid_stream])
```